### ISSUE TITLE
IDE crashes

### ISSUE THREAD

**Issue by jayrulez** [beefytech/Beef#1866]
Here are some IDE crash dumps.
Hopefully they can help to fix some errors.

1 occurs when starting to debug the program.
The other 2 occurs when trying to navigate to the symbol under the cursor by Ctrl+LMB

[Crashes.zip](https://github.com/beefytech/Beef/files/11872895/Crashes.zip)

**Comment by jayrulez** [comment]
https://github.com/beefytech/Beef/assets/78807/3bfe17ef-8e64-4f42-8930-e0b91fecb916

Here's another crash but no stack trace

**Comment by jayrulez** [comment]
[PersistentCrash.zip](https://github.com/beefytech/Beef/files/11889497/PersistentCrash.zip)

@bfiete  here's another. Let me know if these crash dumps are actually useful and whether to keep posting them or not.

**Comment by bfiete** [comment]
Yeah, keep em coming. They are actually usually fairly obvious and easy to fix.

**Comment by bfiete** [comment]
You named your zip "Pesistent Crash". Are these one-time anomalies or something you're persistently (and maybe even reproducibly) seeing?

**Comment by jayrulez** [comment]
> You named your zip "Pesistent Crash". Are these one-time anomalies or something you're persistently (and maybe even reproducibly) seeing?

This one was persistent as long as I didn't change the code in the before re-opening in the IDE. I launched in safe mode and edited the code before the normal mode of the IDE would stop crashing.

**Comment by bfiete** [comment]
If that was code you don't mind sharing then you could have send the project....

**Comment by jayrulez** [comment]
Unfortunately, I did not know exactly which piece of code caused it and I already converted all that code to Beef. It was a bit of C# code I was in the process of converting.

I will see if I can try to reproduce it and share if it crashes again.

**Comment by jayrulez** [comment]
Here are two more crash dumps:
[Crashes.zip](https://github.com/beefytech/Beef/files/11931728/Crashes.zip)


And a project that's presenting one of them.
[Code.zip](https://github.com/beefytech/Beef/files/11931727/Code.zip)

**Comment by jayrulez** [comment]
[CrashJS.dmp](https://github.com/beefytech/Beef/files/12922642/CrashJS.dmp)

Another crash. This is caused by a project that has some javascript files in the Beef project directory.

**Comment by jayrulez** [comment]
[Delegates.zip](https://github.com/beefytech/Beef/files/14156010/Delegates.zip)

Another crash

**Comment by bfiete** [comment]
Maybe this will help b7226145012b9d6de04531cfec33364dad44c98d

**Comment by jayrulez** [comment]
> Repro?

```
struct Event
{
}

delegate void HandleEventCallback<T>(T ev) where T : Event;

struct EventHandle {
	public int Id;
}

struct EventHandle<T> : EventHandle
where T : Event
{
	private HandleEventCallback<T> mCallback;

	public HandleEventCallback<T> Callback => mCallback;

	public this(HandleEventCallback<T> callback)
	{
		mCallback = callback;
	}
}

class EventDispatcher
{
	private Dictionary<Type, List<EventHandle>> mCallbacks = new .();

	private EventHandle<T> AddCallback<T>(HandleEventCallback<T> callback) where T : Event
	{
		var evType = typeof(T);
		if (!mCallbacks.ContainsKey(evType))
			mCallbacks[evType] = new .();

		EventHandle<T> evt = .(callback);

		mCallbacks[evType].Add(evt);

		return evt;
	}

	public EventHandle<T> On<T>(HandleEventCallback<T> callback) where T : Event
	{
		return AddCallback<T>(callback);
	}

	public EventDispatcher Fire<T>(T arg) where T : Event
	{
		var evType = typeof(T);
		if (!mCallbacks.ContainsKey(evType))
			return this;

		var callbacks = mCallbacks[evType];

		for (var evt in callbacks)
		{
			var typedEvt = (EventHandle<T>)*(EventHandle<T>*)(void*)&evt;
			//if(let typedEvt = evt as EvHandle<T>?)
			{
				typedEvt.Callback.Invoke(arg);
			}
		}

		return this;
	}
}
```

try removing arg from this line: typedEvt.Callback.Invoke(arg);
then try to type it back

**Comment by bfiete** [comment]
Fixed at 6302416e403c9b2900ab229e950f2cc3e0f801ec
This was a pretty serious issue with generic constraints on delegates.

It would have been nice to have this issue be a separate GH issue so we could close it and have history of it instead of just being lost in a generic "IDE crashes" item.. but oh well.

**Comment by jayrulez** [comment]
> Fixed at [6302416](https://github.com/beefytech/Beef/commit/6302416e403c9b2900ab229e950f2cc3e0f801ec) This was a pretty serious issue with generic constraints on delegates.
> 
> It would have been nice to have this issue be a separate GH issue so we could close it and have history of it instead of just being lost in a generic "IDE crashes" item.. but oh well.

I mean, it's not too late: https://github.com/beefytech/Beef/issues/1926

**Comment by bfiete** [comment]
That works :-)

**Comment by kallisto56** [comment]
I believe this issue may be closed. Workspace attached in https://github.com/beefytech/Beef/issues/1866#issuecomment-1616983438 no longer crashes the IDE on latest build from source (https://github.com/beefytech/Beef/commit/91483c90af08b6cbb5b41f8902f85f774cab3641). Snippet of code (`EventDispatcher`) also works fine.
